torch/_inductor/freezing.py

```diff
@@ -224,3 +230,36 @@ def discard_traced_gm_params(mod):
             e_t.requires_grad_(True)
             e_t._is_param = True
         setattr(mod, attr_name, e_t)
+
+
+@dynamo_timed
+def convert_conv_weights_to_channels_last(gm):
+    """
+    Convert 4d convolution weight tensor to channels last format.
+
+    This method assumes the graph is already freezed.
+    """
+    convs = [n for n in gm.graph.nodes if n.target == aten.convolution.default]
+    for conv in convs:
+        weight_node = conv.args[1]
+        # is a constant tensor
+        if weight_node.op == "get_attr":
+            param_tensor = getattr(gm, weight_node.target)
+            if len(param_tensor.shape) != 4:
+                # not a 4d tensor, skip
+                continue
+            with no_dispatch():
+                cl_param_tensor = param_tensor.to(memory_format=torch.channels_last)
```

Comment:
Could we convert the tensors to channels last by transformation in the graph ? This is going to compose better with the rest of the freezing stack... In particular, to do real-time weight updates, we are going to need to record all of the transformations we do on weights, and then reapply them. If the channels-last conversion occurs in the graph, it will occur naturally with the other transforms, like mm weight padding

Sorry this is kind of a new wrinkle (real-time updates) I just heard about.

Reply:
No worries. Just to clarify, so you suggest to use a fx.Transformer to do what I'm doing here?

Reply:
We would be inserting `torch.ops.aten.clone.default(weight, memory_format = torch.channels_last)` nodes in the graph on the weights. You could use a fx.Transformer here but i think just iterating over the graph is fine.

Reply:
I see. Do also need to reorder the freezing pass and convert_conv_weights_to_channels_last pass?

Currently convert_conv_weights_to_channels_last happens right after freezing. But if we are add the layout conversion nodes (`aten.clone`) to the graph, we should do convert_conv_weights_to_channels_last before the freezing pass? Otherwise those aten.clone node gonna appear at runtime.

Reply:
Yea, that's correct. We would need to insert it before the freeze call here https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L684, on the aot_autograd graph

Reply:
Wrote up the issue here: https://github.com/pytorch/pytorch/issues/103990